### PR TITLE
Drools query

### PR DESCRIPTION
**Pull request by markbjerke**
This addresses issue #160 concerning the addition of a query that is modeled after the Drools Query facility.

### CODE REVIEW

**Comment by doug-martin** [comment]
Im going to do one pass for formatting, then another for logic. As a side not it you could squash your commits that would be great!

File: lib/compile/index.js [hunk 1]
```diff
@@ -117,12 +117,18 @@ var createRuleFromObject = (function () {
...
+        var name		= obj.name;
+        var options		= obj.options || {};
+        options.scope	= scope;
+
         if (extd.isEmpty(obj)) {
             throw new Error("Rule is empty");
         }
-        var options = obj.options || {};
-        options.scope = scope;
+		if( obj.isQuery ) {
```
Comment (doug-martin) [inline lib/compile/index.js]: Formatting. Please make the indenting match the rest of the block

File: lib/compile/index.js [hunk 2]
```diff
@@ -163,6 +169,8 @@ exports.compile = function (flowObj, options, cb, Container) {
         defined.Buffer = Buffer;
     }
     var scope = merge({console: console}, options.scope);
+	scope.queries = flow.__queries;
```
Comment (doug-martin) [inline lib/compile/index.js]: same

File: lib/compile/index.js [hunk 3]
```diff
@@ -187,7 +195,8 @@ exports.compile = function (flowObj, options, cb, Container) {
     var rules = flowObj.rules;
     if (rules.length) {
         forEach(rules, function (rule) {
-            flow.__rules = flow.__rules.concat(createRuleFromObject(rule, defined, scope));
+			var instance = createRuleFromObject(rule, defined, scope);
```
Comment (doug-martin) [inline lib/compile/index.js]: same

File: lib/constraint.js [hunk 4]
```diff
@@ -225,8 +225,8 @@ Constraint.extend({
             return instanceOf(constraint, this._static) && this.get("alias") === constraint.get("alias") && deepEqual(this.constraints, constraint.constraints);
         },
 
-        "assert": function (fact, fh) {
-            return this.constraints(fact, fh);
+        "assert": function (fact, fh, session) {
+			return this.constraints.call(session, fact, fh);
```
Comment (doug-martin) [inline lib/constraint.js]: formatting.

File: lib/constraint.js [hunk 5]
```diff
@@ -236,7 +236,24 @@ Constraint.extend({
         }
 
     }
-}).as(exports, "FromConstraint");
+});
+exports.FromConstraint = FromConstraint;
+
+var QueryConstraint = FromConstraint.extend({
+    instance: {
+		 constructor: function (constraints, options) {
```
Comment (doug-martin) [inline lib/constraint.js]: formatting.

File: lib/constraintMatcher.js [hunk 6]
```diff
@@ -187,10 +187,59 @@ var lang = {
...
+        var ret = [];
+        var rule2 = rule[2];
+
+        if (rule2 === "identifier") {
+            //its an identifier so stop
+            return [ (rule[0] + identSuffix) ];
+        } else if (rule2 === "function") {
+            ret = ret.concat(this.getArguments(rule[0])).concat(this.getArguments(rule[1]));
+        } else if (
+			rule2 !== "string" &&
```
Comment (doug-martin) [inline lib/constraintMatcher.js]: formatting.

File: lib/constraintMatcher.js [hunk 7]
```diff
@@ -219,35 +268,39 @@ var lang = {
...
-                ret.push(new atoms[isReference ? "ReferenceLTConstraint" : "ComparisonConstraint"](rule, options));
-                break;
-            case "lte":
-                ret.push(new atoms[isReference ? "ReferenceLTEConstraint" : "ComparisonConstraint"](rule, options));
-                break;
-            default:
-                ret.push(new atoms[isReference ? "ReferenceConstraint" : "ComparisonConstraint"](rule, options));
-            }
-
+			if( rule2 === 'function' && queries && queries[rule[0][0]] ) {	// fnName = rule[0][0], it's a query..
```
Comment (doug-martin) [inline lib/constraintMatcher.js]: formatting.

File: lib/constraintMatcher.js [hunk 8]
```diff
@@ -411,16 +464,20 @@ var lang = {
 var matcherCount = 0;
 var toJs = exports.toJs = function (rule, scope, alias, equality, wrap) {
     /*jshint evil:true*/
-    var js = lang.parse(rule);
-    scope = scope || {};
-    var vars = lang.getIdentifiers(rule);
-    var closureVars = ["var indexOf = definedFuncs.indexOf; var hasOwnProperty = Object.prototype.hasOwnProperty;"], funcVars = [];
-    extd(vars).filter(function (v) {
+	var  closureVars = ["var indexOf = definedFuncs.indexOf; var hasOwnProperty = Object.prototype.hasOwnProperty;"],
```
Comment (doug-martin) [inline lib/constraintMatcher.js]: formatting.

**Comment by doug-martin** [comment]
After going through a couple of files I wonder if the spacing is off a little bit in your editor or maybe github is showing the spacing a little weird. I'll let you go through and fix the spacing issue without me cluttering the PR with the same comment.

File: lib/context.js [hunk 9]
```diff
@@ -16,77 +18,28 @@ function createContextHash(paths, hashCode) {
...
-    while (++i < l) {
-        alias = aliases[i];
-        h1[alias] = h2[alias];
-    }
-}
-
-function unionRecency(arr, arr1, arr2) {
-    pPush.apply(arr, arr1);
-    var i = -1, l = arr2.length, val, j = arr.length;
+function createPathstHash(paths, hashCode) {
```
Comment (doug-martin) [inline lib/context.js]: spelling

File: lib/extended.js [hunk 10]
```diff
@@ -127,6 +128,27 @@ function union(arr1, arr2) {
...
+function findIndex(coll, item) {
+	var target = -1;
+	if( 'function' === typeof item ) {
+		coll.some(function(e, i) {
+			if(item(e,i)) {
+				target = i;
+				return true;
+			};
+		});
+	}
```
Comment (doug-martin) [inline lib/extended.js]: Auto format wont catch this most likely but else should be on the same line

File: lib/flow.js [hunk 11]
```diff
@@ -111,7 +123,22 @@ module.exports = declare(EventEmitter, {
...
-        }
+        },
 
+		getQuery: function(name) {
+			return this.__queries[name];
+		},
+
+		getParamNode: function(pattern) {
+			var paramNodes = this.rootNode.paramNodes
+				,target;
```
Comment (doug-martin) [inline lib/flow.js]: comma should be on the previous line.

File: lib/flowContainer.js [hunk 12]
```diff
@@ -42,20 +43,37 @@ var FlowContainer = declare({
...
+			options.scope = options.scope || {};
+			options.scope.queries = this.__queries;
+			instance = rule.createRule(name, options, conditions, cb);
+			this.__addRule(instance);
             return this;
         },
 
+		query: function(name, options, conditions) {
+			var instance, pattern, scope;
+			instance	= rule.createQuery(name, options,conditions);
```
Comment (doug-martin) [inline lib/flowContainer.js]: formatting after the declaration

File: lib/flowContainer.js [hunk 13]
```diff
@@ -70,8 +88,17 @@ var FlowContainer = declare({
...
             });
-        }
+        },
 
+		__addRule: function(ruleInstance) {
+			if( ruleInstance instanceof rule.Query ) {
+				this.__queries[ruleInstance.name] = ruleInstance;
+				this.__rules.push(ruleInstance);						
+			}
+			else {
```
Comment (doug-martin) [inline lib/flowContainer.js]: previous line

File: lib/nodes/ParamNode.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+var FromNode = require("./fromNode"),
+    extd = require("../extended"),
+    Context = require("../context");
+ 
+FromNode.extend({
+    instance: {
+        nodeType: "ParamNode",
+		//
+        constructor: function (pattern, wm) {
```
Comment (doug-martin) [inline lib/nodes/ParamNode.js]: If you are not overriding anything then you can just remove the constructor

File: lib/nodes/collectNode.js [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            this._super([pattern, wm]);
+            this.setAliases			= pattern.setOrientedAliases;
+			this.collectionHash		= {};
+			this.isCollectionObject	= this.type({}) ? true : false
+			this.fnCollectionSrc	= pattern.fnCollectionSrc;
+        },
+
+        __createMatches: function (lc) {
+            var me = this
+                ,fc					= me.getFilteredContext(lc)
```
Comment (doug-martin) [inline lib/nodes/collectNode.js]: commas on the previous line

**Comment by markbjerke** [comment]
Roger that!
-m

File: lib/flowContainer.js (hunk 13 above)
Comment (markbjerke) [inline lib/flowContainer.js]: not sure what you mean here... ?

File: lib/flowContainer.js (hunk 12 above)
Comment (markbjerke) [inline lib/flowContainer.js]: not sure what you mean on this ?

File: lib/extended.js (hunk 10 above)
Comment (markbjerke) [inline lib/extended.js]: OK

File: lib/constraintMatcher.js (hunk 8 above)
Comment (markbjerke) [inline lib/constraintMatcher.js]: not clear to me other than tightening the var decl what the formatting issue is ?

File: lib/flow.js (hunk 11 above)
Comment (markbjerke) [inline lib/flow.js]: OK

**Comment by markbjerke** [comment]
@doug-martin  - I tried to clean up the formatting issues.  The main thing was switched my editor to insert spaces (4) in place of the tab char.

**Comment by markbjerke** [comment]
I am closing this pull request with a plan to reopen it with a smaller delta from the master branch. I decided that there was refactor work included in the PR that should have been broken into separate PR requests.

**Comment by markbjerke** [comment]
I am closing this pull request with a plan to reopen it with a smaller delta from the master branch. I decided that there was refactor work included in the PR that should have been broken into separate PR requests.

**Pull request closed without merging**